src/lib.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! Yatp is a thread pool that tries to be adaptive, responsive and generic.
+
+/// In Yatp model, any pieces of logic aims to be executed in thread pool is
+/// called Task. There can be different definitions of Task. Some people may
+/// choose `Future` as Task, some may just want callbacks, or even Actor
+/// messages. But no matter what a Task is, there should be some role know how
+/// to execute it. The role is call `Runner`.
+/// 
+/// The life cycle of a Runner is:
+/// ```text
+///   start
+///     |
+///     | <--- resume
+///     |        |
+///   handle -> pause
+///     |
+///    end
+/// ```
+/// 
+/// Generally users should use the provided future thread pool or callback
+/// thread pool instead. This is only for advance customization. 
+pub trait Runner {
+    /// The task runner can handle.
+    type Task;
```

Comment:
I once meet some trouble if `Task` knows nothing about the queue (`LocalSpawn` and `RemoteSpawn` here).

For example, a callback handle should be aware of what `LocalSpawn` is, but with this trait it's hard to do that.